internal/docstore/firedocstore/fs.go

```diff
@@ -372,51 +372,53 @@ func (c *collection) newGetRequest(gets []*driver.Action) (*pb.BatchGetDocuments
 	return req, nil
 }
 
+// commitCall holds information needed to make a Commit RPC and to follow up after it is done.
+type commitCall struct {
+	writes   []*pb.Write      // writes to commit
+	actions  []*driver.Action // actions corresponding to those writes
+	newNames []string         // new names for Create; parallel to actions
+}
+
 // runWrites executes all the actions in a single RPC. The actions are done atomically,
 // so either they all succeed or they all fail.
 func (c *collection) runWrites(ctx context.Context, actions []*driver.Action, errs []error, opts *driver.RunActionsOptions) {
-	setErr := func(err error) {
-		for _, a := range actions {
-			errs[a.Index] = err
-		}
-	}
-
 	// Convert each action to one or more writes, collecting names for newly created
-	// documents along the way.
-	var pws []*pb.Write
-	newNames := map[*driver.Action]string{} // from Creates without a name to the new name
+	// documents along the way. Divide writes into those with preconditions and those without.
+	// Writes without preconditions can't fail, so we can execute them all in one Commit RPC.
+	// All other writes must be run as separate Commits.
+	var (
+		nCall  = &commitCall{} // for writes without preconditions
+		pCalls []*commitCall   // for writes with preconditions
+	)
 	for _, a := range actions {
 		ws, nn, err := c.actionToWrites(a)
 		if err != nil {
 			errs[a.Index] = err
-		} else {
-			if nn != "" {
-				newNames[a] = nn
-			}
-			pws = append(pws, ws...)
+		} else if ws[0].CurrentDocument == nil { // no precondition
+			nCall.writes = append(nCall.writes, ws...)
+			nCall.actions = append(nCall.actions, a)
+			nCall.newNames = append(nCall.newNames, nn)
+		} else { // writes have a precondition
+			pCalls = append(pCalls, &commitCall{
+				writes:   ws,
+				actions:  []*driver.Action{a},
+				newNames: []string{nn},
+			})
 		}
 	}
-	// Call the Commit RPC with the list of writes.
-	wrs, err := c.commit(ctx, pws, opts)
-	if err != nil {
-		// TODO(jba): re-run actions so that it appears they happen independently.
-		setErr(err)
-		return
-	}
-	// Now that we've successfully done the action, set the names for newly created docs
-	// that weren't given a name by the caller.
-	for a, nn := range newNames {
-		_ = a.Doc.SetField(c.nameField, nn)
+	// Run the commit calls concurrently.
+	// TODO(jba): limit number of outstanding RPCs.
```

Comment:
We should have an issue to look into doing this at the docstore package level.

Reply:
Filed #2228.